modulus/models/gnn_layers/distributed_graph.py

```diff
@@ -0,0 +1,357 @@
+            self.num_global_src_nodes, src_offsets_in_partition[-1]
+        )
+
+        scatter_indices = [None] * self.partition_size
+        sizes = [
+            [None for _ in range(self.partition_size)] for _ in range(partition_size)
+        ]
+
+        for rank in range(self.partition_size):
+            offset_start = dst_offsets_in_partition[rank]
+            offset_end = dst_offsets_in_partition[rank + 1]
+            offsets = global_offsets[offset_start : offset_end + 1].detach().clone()
+            partition_indices = (
+                global_indices[offsets[0] : offsets[-1]].detach().clone()
+            )
+            offsets -= offsets[0].item()
+
+            global_src_ids_per_rank, inverse_mapping = partition_indices.unique(
+                sorted=True, return_inverse=True
+            )
+            local_src_ids_per_rank = torch.arange(
+                0,
+                global_src_ids_per_rank.size(0),
+                dtype=offsets.dtype,
+                device=offsets.device,
+            )
+            global_src_ids_to_gpu = global_src_ids_per_rank // src_nodes_in_partition
+            remote_src_ids_per_rank = (
+                global_src_ids_per_rank - global_src_ids_to_gpu * src_nodes_in_partition
+            )
+
+            indices = local_src_ids_per_rank[inverse_mapping]
+            self.num_indices_in_each_partition[rank] = indices.size(0)
+
+            if rank == self.partition_rank:
+                self.num_local_indices = indices.size(0)
+                self.num_indices_in_partition = self.num_local_indices
+                self.num_local_dst_nodes = offsets.size(0) - 1
+                self.num_dst_nodes_in_each_partition = [
+                    dst_offsets_in_partition[rank + 1] - dst_offsets_in_partition[rank]
+                    for rank in range(self.partition_size)
+                ]
+                self.num_dst_nodes_in_partition = self.num_local_dst_nodes
+                self.num_local_src_nodes = global_src_ids_per_rank.size(0)
+                self.num_src_nodes_in_each_partition = [
+                    src_offsets_in_partition[rank + 1] - src_offsets_in_partition[rank]
+                    for rank in range(self.partition_size)
+                ]
+                self.num_src_nodes_in_partition = self.num_src_nodes_in_each_partition[
+                    rank
+                ]
+                self.partitioned_src_node_ids_to_global = range(
+                    src_offsets_in_partition[rank], src_offsets_in_partition[rank + 1]
+                )
+                self.partitioned_dst_node_ids_to_global = range(
+                    dst_offsets_in_partition[rank], dst_offsets_in_partition[rank + 1]
+                )
+                self.partitioned_indices_to_global = range(
+                    global_offsets[offset_start], global_offsets[offset_end]
+                )
+
+                self.local_offsets = offsets.detach().clone().to(device=self.device_id)
```

Comment:
I think this might be a leftover of an intermediate version which used things differently and thus this assignment needed a detach/clone as offsets was just a view of a larger global buffer. IIUC, this shouldn't be the case anymore and indeed could go away.